[semi-auto-props]: Add more tests (GetMembers, speculative semantic model)

Youssef1313 opened the pull request:
Test plan: https://github.com/dotnet/roslyn/issues/57012

Youssef1313 commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,36 @@ public S(int arg)
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
+            Assert.Equal(SpecialType.System_Int32, fieldKeywordTypeInfo.Type.SpecialType);
```
Can you please verify if this is expected to work? Currently it doesn't work because (as far as I understand), before speculation, we already marked the backing field as computed.

AlekseyTs replied:
> Can you please verify if this is expected to work? 

Define "to work". 
We are in the investigation territory here. Reviewers should be able to see the current behavior. Meaning, the tests should pass, they should reflect the current behavior. Then we can examine if the behavior matches our expectations.

Youssef1313 commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
+
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
```
This assertion fails. What I was trying to do in this test is to not cause binding to the original compilation so that the sentinel value remains, then I only bind the speculated node. This caused the property to have a backing field, which `GetFieldsToEmit` will return. @AlekseyTs Any pointers as to what I need to do, for both this test and the previous one?

AlekseyTs replied:
> Any pointers as to what I need to do, for both this test and the previous one?

I agree that the behavior doesn't feel right. But before doing anything to fix it, let's add more tests to better understand the set of scenarios we should handle.

Note, the tests should pass, asserts should reflect the current behavior. Otherwise, reviewers cannot "see" the current behavior. If some behavior feel unexpected to you, add a comment.

Here are additional tests that I suggest:
- Speculate with a field identifier within a an accessor that uses the field keyword. 
- Speculate with a field identifier within a an accessor that doesn't use the field keyword, but there is another accessor that uses the keyword.
- Speculate in a duplicate accessor (in the second getter, for example). Both when the first accessor uses and doesn't use the field keyword.

Once we have these tests, I might suggest more scenarios to try.

Youssef1313 replied:
@AlekseyTs For each of those scenarios, do you want two tests:

1. Bind original tree first
2. Bind speculated tree first

as I demonstrated in https://github.com/dotnet/roslyn/pull/59270#discussion_r800035455

If not, which scenario of the two is the interesting one?

AlekseyTs replied:
> For each of those scenarios, do you want two tests?

Yes, let's have two a pair of tests per scenario.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -77,6 +74,9 @@ public static void Main()
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>GetMembers(GeneratedNames.MakeBackingFieldName("P")) [](http://example.com/codeflow?start=57&length=52)

Could we simplify the check to assert that there are no fields? The current check depends on us guessing the "right" name, but that is just another thing to get wrong and fail to detect a wrong behavior.

AlekseyTs replied:
It looks like this suggestion is applicable to other similar checks.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -153,6 +153,7 @@ public class C
 ");
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>MakeBackingFieldName [](http://example.com/codeflow?start=83&length=20)

Same suggestion.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -204,6 +205,7 @@ public class C
 ");
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>MakeBackingFieldName [](http://example.com/codeflow?start=83&length=20)

Same suggestion.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -272,6 +274,7 @@ public class C
 ");
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>MakeBackingFieldName [](http://example.com/codeflow?start=83&length=20)

Same suggestion.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -341,6 +344,7 @@ public class C
 ");
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>MakeBackingFieldName [](http://example.com/codeflow?start=83&length=20)

Same suggestion.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1301,7 +1337,9 @@ void M3()
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
-
+            Assert.NotEmpty(comp.GetTypeByMetadataName("S_WithAutoProperty").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
```
>Assert.NotEmpty(comp.GetTypeByMetadataName("S_WithAutoProperty").GetMembers(GeneratedNames.MakeBackingFieldName("P"))); [](http://example.com/codeflow?start=12&length=119)

Here I would assert the value of ```comp.GetTypeByMetadataName("S_WithAutoProperty").GetMembers().OfType<FieldSymbol>().Single().ToTestDisplayString()``` instead.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
+            var tree = comp.SyntaxTrees[0];
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(comp.SyntaxTrees[0]);
```
>comp.SyntaxTrees[0] [](http://example.com/codeflow?start=46&length=19)

```tree```?

Youssef1313 commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(comp.SyntaxTrees[0]);
+            var identifier = SyntaxFactory.IdentifierName(SyntaxFactory.Identifier(leading: default, contextualKind: SyntaxKind.FieldKeyword, text: "field", valueText: "field", trailing: default));
```
@AlekseyTs `ContextualKind()` will not return `FieldKeyword` if we use `ParseExpression("field")`, so the following will fail 

https://github.com/dotnet/roslyn/blob/f403873b7904fefdca00556690692cfbbdaa55bb/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L1560

AlekseyTs replied:
> ```ContextualKind()``` will not return ```FieldKeyword``` if we use ```ParseExpression("field")```

This feels like a real problem. I would expect a field identifier parsed like that to work as a "field keyword". Why does it work for an identifier within an accessor?

Youssef1313 replied:
@AlekseyTs I'll see if I can understand why this happens.

Youssef1313 replied:
@AlekseyTs This happens in this line:

https://github.com/dotnet/roslyn/blob/baedc51c4cb3840a021bc972a399097f733c0838/src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxFactory.cs#L165

Should I replicate this logic:

https://github.com/dotnet/roslyn/blob/baedc51c4cb3840a021bc972a399097f733c0838/src/Compilers/CSharp/Portable/Parser/LexerCache.cs#L26-L32

followed by a `if (kind == SyntaxKind.None) kind = SyntaxKind.IdentifierToken;` ?

This is pretty much what happens in:

https://github.com/dotnet/roslyn/blob/baedc51c4cb3840a021bc972a399097f733c0838/src/Compilers/CSharp/Portable/Parser/Lexer.cs#L2251-L2259

AlekseyTs replied:
The following test passes for me in 'main'.
```
        [Fact]
        public void Test()
        {
            var id = SyntaxFactory.ParseExpression("field");
            Assert.Equal(SyntaxKind.FieldKeyword, ((IdentifierNameSyntax)id).Identifier.ContextualKind());
        }
```

What am I missing?

Youssef1313 replied:
@AlekseyTs Indeed, my bad. I didn't use `ParseExpression`. `ContextualKind()` doesn't return `FieldKeyword` for `SyntaxFactory.IdentifierName("field")`. But it works with `ParseExpression`.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
+            var model = comp.GetSemanticModel(comp.SyntaxTrees[0]);
+            var identifier = SyntaxFactory.IdentifierName(SyntaxFactory.Identifier(leading: default, contextualKind: SyntaxKind.FieldKeyword, text: "field", valueText: "field", trailing: default));
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
```
>Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit()); [](http://example.com/codeflow?start=12&length=64)

Are you following some specific suggestion by adding this assert? Otherwise, I would remove it because it is going to trigger some work and it is not clear that we want it to be done at this point.

AlekseyTs replied:
Or is it your intent to call it before and call it after the speculation in the next test? Just want to make sure I understand your intent.

Youssef1313 replied:
@AlekseyTs My intent for the two tests was:

- Let one test bind the original tree first (`SpeculativeSemanticModelTest`) then see if there are emitted fields if we speculated with `field`. This was following one of your suggestion in a previous PR, which I recorded as a PROTOTYPE that I should delete in this PR:

    > Speculating with a field keyword within an accessor of a regular property then emitting shouldn't emit the backing field.
    
- Let the other test bind the speculated tree first (`SpeculativeSemanticModelTest2`).

    Since the backing field wasn't marked as calculated, this caused `GetFieldsToEmit()` to return a backing field, which I'm not sure is intended.

Meaning that, currently, the order of binding changes `GetFieldsToEmit()` behavior

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
```
>GetTypeInfo [](http://example.com/codeflow?start=56&length=11)

I think it is better to use ```GetSymbolInfo``` and then assert ```symbol.ToTestDisplayString()``` if we got a symbol back.

AlekseyTs commented:
Done with review pass (commit 3)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,64 @@ public S(int arg)
+            var model = comp.GetSemanticModel(tree);
+            var identifier = SyntaxFactory.IdentifierName(SyntaxFactory.Identifier(leading: default, contextualKind: SyntaxKind.FieldKeyword, text: "field", valueText: "field", trailing: default));
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
```
>GetTypeInfo [](http://example.com/codeflow?start=56&length=11)

Similar to the previous test, I think it is better to use ```GetSymbolInfo``` and then assert ```symbol.ToTestDisplayString()``` if we got a symbol back.

Youssef1313 replied:
Sorry I missed it the first time. Fixed now.

AlekseyTs commented:
Done with review pass (commit 4)

Youssef1313 commented:
@AlekseyTs I added more tests. This is ready for another look.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+        }
+
+        [Fact]
+        public void SpeculativeSemanticModel_FieldInRegularAccessor_BindOriginalFirst()
```
>SpeculativeSemanticModel_FieldInRegularAccessor_BindOriginalFirst [](http://example.com/codeflow?start=20&length=65)

Another scenario that came to mind is Expression Evaluation, see ```Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator.UnitTests.ExpressionCompilerTests.EvaluateThis``` unit-test for example.
I think we need to test an attempt to evaluate ```field``` within an accessor:
- of a regular property (doesn't use `field` keyword)
- of a property that uses `field` keyword.

Also, edit-and-continue (ENC) scenarios are interesting. Specifically when an edit introduces/eliminates usages of `field` keyword for a property. A PROTOTYPE comment for ENC tests should be good enough for this particular PR.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+        [InlineData("get { return field; } get => 0;", 31, true)]
+        [InlineData("get { return 0; } get { return field; }", 27, false)]
+        [InlineData("get { return field; } get { return 0; }", 31, true)]
+        public void SpeculativeSemanticModel_FieldInDuplicateAccessorWhereFirstAccessorUsesField_BindOriginalFirst(string accessors, int diagnosticColumn, bool fieldAccessorFirst)
```
There are other similar scenarios in the list.

Youssef1313 replied:
I'll update the test name to say "OneAccessor" instead of "FirstAccessor"

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+
+            var fieldKeywordSymbolInfo = speculativeModel.GetSymbolInfo(identifier);
+            Assert.Equal("System.Double C.<P>k__BackingField", fieldKeywordSymbolInfo.Symbol.ToTestDisplayString(includeNonNullable: true));
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
```
>0 [](http://example.com/codeflow?start=25&length=1)

Is 0 an expected value for this scenario? Shouldn't ``GetFieldsToEmit()``` call above trigger an accessor binding?

AlekseyTs replied:
I guess ```comp.VerifyDiagnostics();``` call is the reason.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+            var fieldKeywordSymbolInfo = speculativeModel.GetSymbolInfo(identifier);
+            if (fieldAccessorFirst)
+            {
+                Assert.Equal("System.Double C.<P>k__BackingField", fieldKeywordSymbolInfo.Symbol.ToTestDisplayString(includeNonNullable: true));
```
>Assert.Equal("System.Double C.<P>k__BackingField", fieldKeywordSymbolInfo.Symbol.ToTestDisplayString(includeNonNullable: true)); [](http://example.com/codeflow?start=16&length=128)

This is interesting. Somehow we are binding in context of a property accessor. I think we need to understand what exactly semantic model is doing.

Youssef1313 replied:
@AlekseyTs I'm not following here. Since the first accessor is using `field`, the backing field was marked as non-null calculated value. So we get a symbol back when speculating with `field`. Anything wrong with the current behavior?

AlekseyTs replied:
> Anything wrong with the current behavior?

Not so much wrong, but somewhat unexpected. We don't create symbols for duplicate accessors, but in order to get the field symbol we need to bind in context of an accessor. So, it feels like we are binding in context of an accessor that has different body. That is probably OK fall back strategy, but we need to confirm/understand exactly what is happening, whether that is an intentional fall back strategy or just an accident, etc. For example, does it mean that locals from the other accessor are in scope? I am not sure if we want this behavior.

Youssef1313 replied:
@AlekseyTs Thanks for clarifying. I'll investigate this tomorrow, it's a bit late here now :)

Youssef1313 replied:
Let me know if this behavior is okay so I can remove the PROTOTYPE comment.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+        [InlineData("get { return field; } get => 0;")]
+        [InlineData("get { return 0; } get { return field; }")]
+        [InlineData("get { return field; } get { return 0; }")]
+        public void SpeculativeSemanticModel_FieldInDuplicateAccessorWhereFirstAccessorUsesField_BindSpeculatedFirst(string accessors)
```
>FieldInDuplicateAccessorWhereFirstAccessorUsesField [](http://example.com/codeflow?start=45&length=51)

> FieldInDuplicateAccessorWhereFirstAccessorUsesField

Same comment as for the test above.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+        [InlineData("get => 0; get { return 1; }", 19, 1)]
+        [InlineData("get { return 0; } get { return 1; }", 27, 0)]
+        [InlineData("get { return 0; } get { return 1; }", 27, 1)]
+        public void SpeculativeSemanticModel_FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField_BindOriginalFirst(string accessors, int diagnosticColumn, int numericLiteralToSpeculate)
```
>FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField [](http://example.com/codeflow?start=45&length=57)

> FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField

Is this accurate? Is ```[InlineData("get => 0; get => 1;", 19, 0)]``` really binds within a duplicate accessor?

AlekseyTs replied:
> I don't see what's wrong with the new suggested name. Can you elaborate please?

"**One**DoesNotUseField", but it looks like `field` keyword isn't used anywhere.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+        [InlineData("get => 0; get { return 1; }", 1)]
+        [InlineData("get { return 0; } get { return 1; }", 0)]
+        [InlineData("get { return 0; } get { return 1; }", 1)]
+        public void SpeculativeSemanticModel_FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField_BindSpeculatedFirst(string accessors, int numericLiteralToSpeculate)
```
>FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField [](http://example.com/codeflow?start=45&length=57)

> FieldInDuplicateAccessorWhereFirstAccessorDoesNotUseField

Similar comment as for the previous test.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1378,394 @@ public S(int arg)
+            Assert.Equal("System.Double C.<P>k__BackingField", fieldKeywordSymbolInfo.Symbol.ToTestDisplayString(includeNonNullable: true));
             Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
         }
     }
```
>} [](http://example.com/codeflow?start=4&length=1)

I think it will be useful to add similar speculation tests replacing body of accessors that don't use `field` keyword with bodies that use "field" identifier. For example, replace what used to be
```
{
    return 0;
}
```

with 
```
{
    in field = 0;
    return field;
}
```

AlekseyTs replied:
I am still tracking this thread.

Youssef1313 replied:
@AlekseyTs Sorry for missing this comment earlier. I added some compiler tests for this. For expression evaluator, I don't know how can I specify at which point the evaluation happens? For example:

```csharp
    if (GetBoolValue())
    {
        int field = 10;
        return field; // evaluating here should use the local
     }

    return field; // Evaluating here should use the backing field.
```

cston replied:
> For expression evaluator, I don't know how can I specify at which point the evaluation happens? 

The `ExpressionCompilerTests.Evaluate()` helper method has an `atLineNumber:` parameter. There are several tests that use that, typically by including a `#line` directive in the source.

Youssef1313 replied:
Thank you @cston for guidance! I added the tests.

AlekseyTs commented:
Done with review pass (commit 5)

Youssef1313 commented:
@AlekseyTs I addressed the review comments except for https://github.com/dotnet/roslyn/pull/59270#discussion_r803864781 which I'll do tomorrow. Thanks!

Youssef1313 commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs:
```diff
@@ -6952,5 +6952,54 @@ .maxstack  2
+                var testData = new CompilationTestData();
+                var result = context.CompileExpression("field", out var error, testData);
+
+                // PROTOTYPE(semi-auto-props): Likely incorrect behavior.
```
@AlekseyTs Can you confirm please?

Youssef1313 replied:
So looking into this, it's failing because we get a `PEMethod` while binder checks for `ContainingMemberOrLambda is SourcePropertyAccessorSymbol`. I can add a check for the `PEMethod` as well, but I don't see anywhere the binder do that, so I'm feeling this is not the right way to do things. Any pointers @AlekseyTs ? Thanks!

Youssef1313 replied:
I attempted to fix it in 86b7d89, not sure it's the right way to do it.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1380,484 @@ public S(int arg)
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, (IdentifierNameSyntax)identifier, out var speculativeModel);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+
+            var fieldKeywordSymbolInfo = speculativeModel.GetSymbolInfo(identifier);
```
>speculativeModel.GetSymbolInfo(identifier) [](http://example.com/codeflow?start=41&length=42)

Let's also test behavior of GetSpeculativeSymbolInfo API. This comment is applicable to other similar tests as well

Youssef1313 commented:
@AlekseyTs Is the next step to address the cases where we don't like the behavior of speculative semantic model? Are all of them covered by PROTOTYPE comments or do you see cases that're not (or vice versa)? Thanks!

AlekseyTs commented:
>  Is the next step to address the cases where we don't like the behavior of speculative semantic model? Are all of them covered by PROTOTYPE comments or do you see cases that're not (or vice versa)?

I didn't have a chance to complete review of changes made post commit 5. We might want to add tests for more scenarios. Once we'll have the tests in place, we will have to decide what exact behavior do we want and adjust implementation accordingly.

---
In reply to: [1039641573](https://github.com/dotnet/roslyn/pull/59270#issuecomment-1039641573) [](http://example.com/codeflow?ancestors=1039641573,1039641573)

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs:
```diff
@@ -6952,5 +6952,62 @@ .maxstack  2
+            {
+                var context = CreateMethodContext(runtime, "C.get_P");
+                var testData = new CompilationTestData();
+                var result = context.CompileExpression("field", out var error, testData);
```
>field [](http://example.com/codeflow?start=56&length=5)

Let's add a test for a more complex expression. For example, "field + 1".

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs:
```diff
@@ -597,13 +597,22 @@ private EEMethodSymbol GetTypeVariablesMethod(EENamedTypeSymbol container, strin
-                ? binder.BindDeconstruction((AssignmentExpressionSyntax)syntax, bindingDiagnostics, resultIsUsedOverride: true)
-                : binder.BindRValueWithoutTargetType(syntax, bindingDiagnostics);
+            BoundExpression? expression;
+            if (TryBindFieldKeyword(binder.ContainingMemberOrLambda, syntax, out var boundField))
```
>var [](http://example.com/codeflow?start=81&length=3)

Please spell out the type

AlekseyTs commented on src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs:
```diff
@@ -660,6 +669,49 @@ private EEMethodSymbol GetTypeVariablesMethod(EENamedTypeSymbol container, strin
             return new BoundReturnStatement(syntax, RefKind.None, expression) { WasCompilerGenerated = true };
         }
 
+        private static bool TryBindFieldKeyword(Symbol? containingMemberOrLambda, ExpressionSyntax syntax, [NotNullWhen(true)] out BoundFieldAccess? boundField)
```
>TryBindFieldKeyword [](http://example.com/codeflow?start=28&length=19)

This doesn't feel like the right way to address the problem:
- It only handles standalone identifier.
- The identifier might actually refer to a local in scope even when there is a backing field.
- Simply relying on name feels fragile. 

Perhaps the field in metadata should be specially marked (with an attribute?) and the binder chain should be adjusted to bring it in scope at the right time. However, as I already suggested earlier, let's not try to fix/change behavior just yet. Let's focus on adding tests and identifying questionable/unexpected behavior.

AlekseyTs commented:
Done with review pass (commit 10)

AlekseyTs commented:
Done with review pass (commit 11). I didn't see any response to https://github.com/dotnet/roslyn/pull/59270#discussion_r807017949.

Youssef1313 commented:
> Done with review pass (commit 11). I didn't see any response to [#59270 (comment)](https://github.com/dotnet/roslyn/pull/59270#discussion_r807017949).

Sorry for missing this comment. It's partially addressed with a question regarding expression evaluator tests.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -392,13 +397,115 @@ public class C
+                //     public int P { get { int field = field; return field; } }
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "field").WithArguments("field").WithLocation(4, 38)
+            );
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
```
>Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding); [](http://example.com/codeflow?start=12&length=70)

Consider also checking GetFieldsToEmit before this assert.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+            {
+                Assert.Equal("System.Int32 field", fieldKeywordSymbolInfo.Symbol.GetSymbol().ToTestDisplayString());
+                Assert.Equal(SymbolKind.Local, fieldKeywordSymbolInfo.Symbol.Kind);
+                Assert.Equal(2, accessorBindingData.NumberOfPerformedAccessorBinding);
```
>Assert.Equal(2, accessorBindingData.NumberOfPerformedAccessorBinding); [](http://example.com/codeflow?start=16&length=70)

I think it would be good to determine what triggers the binding (by asserting ```NumberOfPerformedAccessorBinding``` in right places), the ```GetFieldsToEmit``` above or speculative binding.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+        [InlineData("get { return field; } get => 0;")]
+        [InlineData("get { return 0; } get { return field; }")]
+        [InlineData("get { return field; } get { return 0; }")]
+        [InlineData("get { return field; } get { int field = 0; return field; }")]
```
>[InlineData("get { return field; } get { int field = 0; return field; }")] [](http://example.com/codeflow?start=8&length=74)

Is testing different order not interesting?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+        [InlineData("get { return 0; } get { return 1; }", 27, 0, false)]
+        [InlineData("get { return 0; } get { return 1; }", 27, 1, false)]
+        [InlineData("get { double field = 0; return field; } get { return 1; }", 49, 0, true)]
+        [InlineData("get { double field = 0; return field; } get { return 1; }", 49, 1, false)]
```
>false [](http://example.com/codeflow?start=88&length=5)

We had an internal discussion about expected behavior for speculative binding. We settled on: "`field` in the speculative model does not bind to a backing field if the original location was not a semi-auto property." This is captured as a check box in [https://github.com/dotnet/roslyn/issues/57012](https://github.com/dotnet/roslyn/issues/57012). It might be worth adding a PROTOTYPE comment here because the behavior doesn't match the expectation.

Youssef1313 replied:
> "`field` in the speculative model does not bind to a backing field **if the original location was not a semi-auto property.**"

@AlekseyTs In case the speculated tree was bound first, we don't yet know if the original is semi-auto property or not. What should we do in that case?

Youssef1313 replied:
> It might be worth adding a PROTOTYPE comment here because the behavior doesn't match the expectation.

@AlekseyTs Taking a look at this specific test again, it seems that we already don't bind to a backing field (`Assert.Null(fieldKeywordSymbolInfo.Symbol);`). Was your comment intended for another test? Am I missing something?

AlekseyTs replied:
> Taking a look at this specific test again, it seems that we already don't bind to a backing field (```Assert.Null(fieldKeywordSymbolInfo.Symbol);```). Was your comment intended for another test? Am I missing something?

I probably misinterpreted the meaning of the ```speculatingProducesLocal``` parameter. Assumed that `false` means field rather than not found.

AlekseyTs replied:
> In case the speculated tree was bound first, we don't yet know if the original is semi-auto property or not. What should we do in that case?

One way or the other we probably need to figure out if the original property is a semi-auto property. Just need to find a "right" way to do that. Always force binding the accessors in the process of speculation obviously is going to work, but we should consider if there are "better" alternatives.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+        [InlineData("get { return 0; } get { return 1; }", 0, false)]
+        [InlineData("get { return 0; } get { return 1; }", 1, false)]
+        [InlineData("get { int field = 0; return field; } get { return 1; }", 0, true)]
+        [InlineData("get { int field = 0; return field; } get { return 1; }", 1, false)]
```
>false [](http://example.com/codeflow?start=81&length=5)

Similar comment as for the previous test.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+                Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+                Assert.Equal("System.Int32 field", fieldKeywordSymbolInfo.Symbol.GetSymbol().ToTestDisplayString());
+                Assert.Equal(SymbolKind.Local, fieldKeywordSymbolInfo.Symbol.Kind);
+                Assert.Equal(1, accessorBindingData.NumberOfPerformedAccessorBinding);
```
>NumberOfPerformedAccessorBinding [](http://example.com/codeflow?start=52&length=32)

Consider also asserting the value before the ```Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());``` above.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:
```diff
@@ -1350,9 +1481,642 @@ public S(int arg)
+            var fieldKeywordSymbolInfo2 = model.GetSpeculativeSymbolInfo(token.SpanStart, identifier, SpeculativeBindingOption.BindAsExpression);
+            Assert.Equal(fieldKeywordSymbolInfo, fieldKeywordSymbolInfo2);
+
+            Assert.Equal("System.Int32 C.<P>k__BackingField", comp.GetTypeByMetadataName("C").GetFieldsToEmit().Single().ToTestDisplayString());
```
>Assert.Equal("System.Int32 C.<P>k__BackingField", comp.GetTypeByMetadataName("C").GetFieldsToEmit().Single().ToTestDisplayString()); [](http://example.com/codeflow?start=12&length=132)

I would expect to see a PROTOTYPE comment because the behavior is not consistent with the one observed in the previous test.

AlekseyTs commented:
Done with review pass (commit 13)

AlekseyTs reviewed (approved):
LGTM (commit 14)

AlekseyTs commented:
@Youssef1313 Thanks for the contribution.

Youssef1313 commented:
Thanks for reviewing @AlekseyTs! What do you think the next step should be? A new scenario? Fixing semantic model tests? something else?

333fred commented:
@Youssef1313 I'll take look over the next day or so and let you know.
